### ISSUE TITLE
Expectations for predict, predict_proba, etc.

### ISSUE THREAD

**Issue by romanlutz** [microsoft/fairlearn#65]
Since fairlearn uses randomized classifiers we need to set expectations for what `predict`, `predict_proba` etc. return.

**Comment by romanlutz** [comment]
API documentation exists, but it doesn't reflect randomization well yet.

**Comment by harvineet** [comment]
For ExponentiatedGradient, what is the preferred way of getting predicted **probabilities**?
Can we use `_best_classifier`?

The notebook example [link](https://github.com/fairlearn/fairlearn/blob/master/notebooks/Mitigating%20Disparities%20in%20Ranking%20from%20Binary%20Data.ipynb) expects `predict` to return scores for `LogisticRegression`:

`scores_expgrad_X = pd.Series(expgrad_X.predict(X_test), name="scores_expgrad_X")`

However, the result seems to be the predicted class.

**Comment by romanlutz** [comment]
Right. For probabilities we currently provide `_pmf_predict` from https://github.com/fairlearn/fairlearn/blob/master/fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py but as you can tell from the underscore there's no guarantee it'll stay that way.

**Comment by harvineet** [comment]
I looked at `_pmf_predict` in Exponentiated Gradient. I have a question on how the predictions from the returned classifiers are combined to get predicted probabilities.

**Current**
In `_mean_pred`[function](https://github.com/fairlearn/fairlearn/blob/master/fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py#L17), the approach taken is to  call `predict` and then weight the result. This does not pass on the 'full uncertainty information' from the classifiers. As `predict` only returns 0 or 1 instead of probabilities. Encountered this in an example where the `_classifiers` had only one classifier. Then, the result from `_pmf_predict` had only 0s and 1s which is what `predict` returns. This is an issue.

**Alternative**
Instead, one can call `predict_proba` for each classifier and then weight. In expectation, the results should be same i.e. expectation of the two variables `weight*Bernoulli(prob)` and `Bernoulli(weight*prob)` is `weight*prob`.

Is this a better way for getting predicted probabilities? Trying it on the [notebook example](https://github.com/fairlearn/fairlearn/blob/master/notebooks/Mitigating%20Disparities%20in%20Ranking%20from%20Binary%20Data.ipynb) seem to give better AUC and disparity values. Did not do a thorough test.

```
def weighted_score(dataset, algorithm):
    pred_all = np.zeros((dataset.shape[0],len(algorithm._classifiers)))
    for i,clf in enumerate(algorithm._classifiers):
        pred_all[:,i] = clf.predict_proba(dataset)[:,1]

    score = pred_all.dot(algorithm._expgrad_result._weights)

    return score

scores_expgrad_X = pd.Series(weighted_score(X_test, expgrad_X), name="scores_expgrad_X")
```

**Comment by romanlutz** [comment]
@MiroDudik what do you think of @harvineet 's idea above? I totally missed the comment... apologies.

**Comment by jiajiexiao** [comment]
Hi all, I have a similar question and found super interesting discussions here. Compared to the class than the predicted label, I'd be more interested in the probability of classifying a sample and hope `ExponentiatedGradient` can return the full uncertainty estimation. 

What @harvineet said above [[link](https://github.com/fairlearn/fairlearn/issues/65#issuecomment-579967910)] seems to me a reasonable approach to achieve my goal.  Meanwhile, I also saw another way to derive `pred_prob` as previously suggested by @MiroDudik in this [thread](https://github.com/Trusted-AI/AIF360/pull/215#discussion_r524511326). 

Given the time differences of those threads, I was wondering what would be the currently recommended way to implement the `predict_proba` method for `ExponentiatedGradient` to get a probabilistic estimate after the fairness mitigation. 

Many thanks in advance.

**Comment by TamaraAtanasoska** [comment]
as part of the efforts going through all the issue backlog, pinging @romanlutz and @MiroDudik again for some thoughts, so we could make a more actionable issue that folks that can take.

**Comment by romanlutz** [comment]
From my perspective, we just need to document well that the outputs are probabilistic. This trips people up because they don't expect it. (Perhaps that's more intuitive with gen AI now haha).

**Comment by TamaraAtanasoska** [comment]
@romanlutz got it, noting :)
